### PR TITLE
Display a warning whenever an menu item's alias has changed

### PR DESCRIPTION
**Pull request by marrouchi**
Display a warning message whenever a menu item's alias has been updated. The message suggests to the admin user to create a new redirection to the old alias in order to avoid 404 errors from existing urls.

Pull Request for Issue #13851.

### Summary of Changes
1. Added two methods to the redirect plugin: 
- onContentbeforeSave : Check if the menu item's alias has changed. If it's the case, then display a warning message that suggests to the admin user to create a redirection to the old alias.
- onContentPrepareForm : Alter menu item edit form whenever there's GET params (old_alias & new_alias) in order to set default values for the fields `old_url` & `new_url`
2. An enable/disable field (`Watch Alias`) has been added to the plugin config form if there's need a disable this feature.

### Testing Instructions
First, make sure that the redirect plugin is enabled.

When enabled :
1. Edit an existing menu item and change the alias. Then hit the Save button.
2. A warning message should be displayed that says : "You should create a new redirection to avoid 404 errors that could be generated by the old alias's urls : **Create a new redirection**"
3. Click on the link **Create a new redirection** and you should be redirected to a new redirection form with the two fields `old_url` & `new_url` filled with default values. Hit the Save button.
4. Access the old URL in frontend (that uses the old alias) and you should be redirected to the new one.

When disabled :
1. Go the redirect plugin config form and disable `Watch Alias`
2. Back to the menu manager,  Edit an existing menu item and change the alias. Then hit the Save button.
3. There will be no warning message.

### Documentation Changes Required
No.

### CODE REVIEW

**Comment by tonypartridge** [comment]
@marrouchi great minds think a like ;-) I just created an issue today about this

issue #13851 

I was thinking of going one step further and auto-creating the redirect if a user wishes. I'll run some tests on this.

**Comment by marrouchi** [comment]
@tonypartridge Yes i saw your issue and i've coded a PR right away :)

**Comment by tonypartridge** [comment]
I am just wondering if a plugin is needed. Can we not just adjust the com_menus component? Since the plugin is called on every save when it is only needed on a menu item save.

**Comment by marrouchi** [comment]
I thought of this approach would be better to keep the menu item controller clean, since this feature is related to two components (com_menus & com_redirect).

**Comment by tonypartridge** [comment]
I see your point completely, but it's then a plugin which is also then still requires both components. But both components ship with the core.

@zero-24 @mbabker what are your thoughts on this?

I was thinking of checking if com_redirects plugin is enabled and if so basically doing what Marrouchi has done. But within the component controller opposed to a plugin?

**Comment by mbabker** [comment]
Why not throw it in the existing redirect plugin?  The notice seems like a good idea to me and that plugin already has to be active to make use of the system, so...

**Comment by tonypartridge** [comment]
Duhh good idea! @mbabker 

@marrouchi do you want to do the changes?

**Comment by marrouchi** [comment]
Yes good idead ! Should i create a new PR ?

**Comment by tonypartridge** [comment]
I don't think so? Just remove this plugin from your GIT and add the changes to the system and push your changes they should come through for us to review again.

**Comment by bertmert** [comment]
One should only get a message or action link (whatever) if sef is enabled. 
And there should be an option in the plugin to deactivate this feature completely, please. Because of other extensions that also deal with aliases.

**Comment by tonypartridge** [comment]
I disagree re see as some people might disable it on testing environments when doing something or other. But an option to disable it is a good idea.

**Comment by marrouchi** [comment]
Done ! I've added a enable/disable button of this features in the redirect plugin config form.

Ready to test :+1:

**Comment by franz-wohlkoenig** [comment]
I have tested this item :red_circle: unsuccessfully on 429b80da162cdb9d5de43b65c2ee786efa5cf667<br /><br />Point 1-3 of Testing Instructions works.

Not working Point 4: Using old Url get redirected on Homepage. Using modern Router.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by marrouchi** [comment]
@franz-wohlkoenig This PR does not alter the redirection mechanism. Redirection does not seem to work when modern routing is enabled and that's an new issue that must be addressed. To confirm this, please revert the patch and try a redirection when having modern routing enabled.

**Comment by franz-wohlkoenig** [comment]
Test on legacy Router, Plugin "Redirect" enabled.

Got on Point 4 (using old alias `using-joomla`) a `404 Category not found`. Category is enabled, in Plugin is a new, disabled Link `index.php/using-joomla` without new URL.

#### Test on:
Joomla! 3.7.0-beta1
macOS Sierra, 10.12.3
Firefox 50.1.0
PHP 7.0.4
MySQLi 5.5.53-0

**Comment by brianteeman** [comment]
Can you look at resolving the conflicts so that this can be tested please.

**Comment by raoulyemetio** [comment]
I have tested this item :red_circle: unsuccessfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<br /><br />I apply the patch, watch alias was enabled
then i changed an existing menu item
but i got no warning message
@icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by coolcat-creations** [comment]
I have tested this item :red_circle: unsuccessfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<br /><br />I would love to have this feature, but i had an error on testing. 

0 Using $this when not in object context
/Applications/MAMP/htdocs/joomla-cms/libraries/src/Application/CMSApplication.php:369 

@Yemetio if you restest later open the plugin and save the settings<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by fatmaakay** [comment]
I have tested this item :white_check_mark: successfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by fatmaakay** [comment]
I have tested this item :white_check_mark: successfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by durubayram** [comment]
I have tested this item :white_check_mark: successfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests considering 2 unsuccessful tests so Maintainers have to decide (as always).

File: plugins/system/redirect/redirect.php [hunk 1]
```diff
@@ -334,4 +334,70 @@ private static function doErrorHandling($error)
...
+	 * @return  boolean   true if function not enabled, is in frontend or is new. Else true or
+	 *                    false depending on success of save function.
+	 *
+	 * @since   1.6
+	 */
+	public function onContentbeforeSave($context, $menuItem, $isNew)
+	{
+		if (!$isNew && $context == 'com_menus.item' && $this->params->get('watch_alias'))
+		{
+			$menu = & JSite::getMenu();
```
Comment (coolcat-creations) [inline plugins/system/redirect/redirect.php:356]: Please change this line to 

			$menu = (new \Joomla\CMS\Application\SiteApplication)->getMenu();

to make the PR work.

**Comment by durubayram** [comment]
I have tested this item :red_circle: unsuccessfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by brianteeman** [comment]
please remove the rtc status

**Comment by Quy** [comment]
Removed RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by fatmaakay** [comment]
I have tested this item :red_circle: unsuccessfully on b9ac171feddd4fcb2b152c8d5a3297da98a2686d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13856">issues.joomla.org/tracker/joomla-cms/13856</a>.</sub>

**Comment by coolcat-creations** [comment]
@marrouchi 
Can you change the Line I reviewed, and fix the Code Style for 3.8 ?
+ Would be great to have this watcher also on trashing and on unpublishing the item.
Thank you for this great addition.

File: plugins/system/redirect/redirect.php [hunk 2]
```diff
@@ -334,4 +334,70 @@ private static function doErrorHandling($error)
...
+	 * to the new one.
+	 *
+	 * @param   string   $context   The context of the content passed to the plugin (added in 1.6)
+	 * @param   object   $menuItem  A JTableContent object
+	 * @param   boolean  $isNew     If the content is just about to be created
+	 *
+	 * @return  boolean   true if function not enabled, is in frontend or is new. Else true or
+	 *                    false depending on success of save function.
+	 *
+	 * @since   1.6
```
Comment (Quy) [inline plugins/system/redirect/redirect.php:350]: Change to `__DEPLOY_VERSION__`

File: plugins/system/redirect/redirect.php [hunk 3]
```diff
@@ -334,4 +334,70 @@ private static function doErrorHandling($error)
...
+
+	/**
+	 * Adds default values to the redirection editing form
+	 *
+	 * @param   JForm  $form  The form to be altered.
+	 * @param   mixed  $data  The associated data for the form.
+	 *
+	 * @return  boolean
+	 *
+	 * @since   1.6
```
Comment (Quy) [inline plugins/system/redirect/redirect.php:378]: Change to `__DEPLOY_VERSION__`

File: plugins/system/redirect/redirect.php [hunk 4]
```diff
@@ -334,4 +334,70 @@ private static function doErrorHandling($error)
...
+	 */
+	public function onContentPrepareForm($form, $data)
+	{
+		if (!($form instanceof JForm))
+		{
+			$this->_subject->setError('JERROR_NOT_A_FORM');
```
Comment (Quy) [inline plugins/system/redirect/redirect.php:384]: Missing language string

**Comment by HLeithner** [comment]
Hi @marrouchi,

as this is a new feature it will not go into J3.x series because of the feature freeze. Also it seams it doesn't work (4 failing tests).
If you still interested to bring this feature into Joomla please rebase the PR on J4 branch and correct the reported errors.

Thx for your work and time you invested into Joomla! I'm closing this PR and hope you find time to create a new one for J4.

**Pull request closed without merging**